### PR TITLE
SM PSCI v1.1 update + iMX WDOG updates

### PR DESCRIPTION
**Pull request by igoropaniuk**
This PR adds support for PSCI v1.1 + reset2 implementation for iMX SoCs:

### CODE REVIEW

File: core/drivers/imx_wdog.c [hunk 1]
```diff
@@ -39,15 +39,15 @@
...
+static bool ext_reset_output;
 static vaddr_t wdog_base;
 
 static const char * const dt_wdog_match_table[] = {
 	"fsl,imx21-wdt",
 	"fsl,imx7ulp-wdt",
 };
 
-void imx_wdog_restart(void)
+void imx_wdog_restart(bool external_reset __unused)
```
Comment (jenswi-linaro) [inline core/drivers/imx_wdog.c:50]: Please use `__maybe_unused` instead.

**Review by jenswi-linaro — COMMENTED** [review]
Comment for the commit "core: arm: sm: replace function ids with hex values"
If we're concerned about matching with the spec then I guess we should use the full id instead of using helper macros.

**Comment by igoropaniuk** [comment]
@jenswi-linaro Personally I do like current approach with PSCI_FN_BASE, as basically we can define it do different value  if someone decides to extend existing SM implementation for aarch64 (having `0xc4000000` as PSCI_FN_BASE), instead of introducing duplicates for every PSCI func (like it's currently done in TF-A). But I can change it to full id if it's preferable.

@clementfaure Could you please take a quick look at these patches?

**Comment by clementfaure** [comment]
For ```imx: pm: add support for reset2 function```
```Acked-by: Clement Faure <clement.faure@nxp.com>```

**Comment by igoropaniuk** [comment]
@jenswi-linaro just let me know which one you prefer.
I can also just drop that patch

**Comment by jenswi-linaro** [comment]
I've found that defining the smc function ids as full ids instead of using helper macros which hides the final value more useful. If you see a function id in a log or something grepping for it in the source can be quite convenient. Also the specifications tends to specify the full id so it's easier when reviewing too. If we're going to touch all the smc function ids in `psci.h` then we should make them full ids.

**Comment by igoropaniuk** [comment]
@jenswi-linaro I've changed to full ids, thanks

File: core/arch/arm/include/sm/psci.h [hunk 2]
```diff
@@ -23,8 +24,11 @@
 #define PSCI_PSCI_SET_SUSPEND_MODE	U(0x8400000f)
 #define PSCI_FN_STAT_RESIDENCY		U(0x84000010)
 #define PSCI_FN_STAT_COUNT		U(0x84000011)
+#define PSCI_SYSTEM_RESET2		U(0x84000012)
+#define PSCI_MEM_PROTECT		U(0x84000013)
+#define PSCI_MEM_CHK_RANGE		U(0x84000014)
```
Comment (jenswi-linaro) [inline core/arch/arm/include/sm/psci.h:29]: I'd prefer to just stick the `PSCI_` prefix to the name in the spec, that is, `PSCI_MEM_PROTECT_CHECK_RANGE`

File: core/arch/arm/sm/psci.c [hunk 3]
```diff
@@ -95,6 +95,23 @@ __weak int psci_features(uint32_t psci_fid __unused)
 	return PSCI_RET_NOT_SUPPORTED;
 }
 
+__weak int psci_mem_protect(uint32_t enable __unused)
+{
+	return PSCI_RET_NOT_SUPPORTED;
+}
+
+__weak int psci_mem_chk_range(uintptr_t base __unused,
+			      uint32_t length __unused)
```
Comment (jenswi-linaro) [inline core/arch/arm/sm/psci.c:104]: `length` should use the `size_t` type.

File: core/arch/arm/sm/psci.c [hunk 4]
```diff
@@ -95,6 +95,23 @@ __weak int psci_features(uint32_t psci_fid __unused)
 	return PSCI_RET_NOT_SUPPORTED;
 }
 
+__weak int psci_mem_protect(uint32_t enable __unused)
+{
+	return PSCI_RET_NOT_SUPPORTED;
+}
+
+__weak int psci_mem_chk_range(uintptr_t base __unused,
```
Comment (jenswi-linaro) [inline core/arch/arm/sm/psci.c:103]: I guess this is physical address so please use `paddr_t` instead.

**Review by jenswi-linaro — COMMENTED** [review]
Comments for "core: arm: sm: upgrade to PSCI 1.1"

**Comment by jenswi-linaro** [comment]
For "core: arm: sm: replace function ids with hex values" please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/include/sm/psci.h (hunk 2 above)
Comment (igoropaniuk) [reply core/arch/arm/include/sm/psci.h]: fixed

**Pull request merged**